src/app/CommandHandler.cpp

```diff
@@ -239,9 +260,45 @@ Status CommandHandler::ProcessInvokeRequest(System::PacketBufferHandle && payloa
 CHIP_ERROR CommandHandler::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PayloadHeader & aPayloadHeader,
                                              System::PacketBufferHandle && aPayload)
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (mState == State::AwaitingResponse && aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        ReturnErrorOnFailure(StatusResponse::ProcessStatusResponse(std::move(aPayload), err));
+        SuccessOrExit(err);
+        ReturnErrorOnFailure(SendCommandResponse());
+        if (mState != State::AwaitingResponse)
+        {
+            // We are sending out the last message and no longer are expecting any responses. As a results we are
+            // fulfilling out responsibility to call close by calling ExitNow().
+            ExitNow();
+        }
+        return CHIP_NO_ERROR;
+    }
     ChipLogDetail(DataManagement, "CommandHandler: Unexpected message type %d", aPayloadHeader.GetMessageType());
     StatusResponse::Send(Status::InvalidAction, mExchangeCtx.Get(), false /*aExpectResponse*/);
-    return CHIP_ERROR_INVALID_MESSAGE_TYPE;
+    err = CHIP_ERROR_INVALID_MESSAGE_TYPE;
+    if (mState == State::AwaitingResponse)
+    {
+        // We were waiting on a response and we got something we did not expect, so we will no longer continue sending
+        // messages, nor do we expect to get any more responses. As a result we are simply fulfilling our
+        // responsibility to call Close() by calling ExitNow().
```

Comment:
Will this cause the right chain of callbacks to occur for the sender?

Reply:
We got a message that we don't understand from the otherside, we are sending out `StatusResponse::Send(Status::InvalidAction,...)` on line 279. I don't thjink there is anything more we can do

Reply:
The question is restated as: when you have sent the InvalidAction, will the CommandSender on the other side clean-up all pending command properly and call the right callbacks? Will it call the non path-specific callback, or the path specific callback each time with the error?

Reply:
It will call the non-path error callback

Reply:
Assuming since you approved after my response here without a follow up comment that this concern has been resolved